src/core/vector/qgsvectorlayereditutils.cpp

```diff
@@ -542,41 +542,52 @@ Qgis::GeometryOperationResult QgsVectorLayerEditUtils::splitParts( const QgsPoin
   QgsFeature feat;
   while ( fit.nextFeature( feat ) )
   {
-    QVector<QgsGeometry> newGeometries;
-    QgsPointSequence topologyTestPoints;
     QgsGeometry featureGeom = feat.geometry();
-    splitFunctionReturn = featureGeom.splitGeometry( splitLine, newGeometries, topologicalEditing, topologyTestPoints, false );
 
-    if ( splitFunctionReturn == Qgis::GeometryOperationResult::Success && !newGeometries.isEmpty() )
+    const QVector<QgsGeometry> geomCollection = featureGeom.asGeometryCollection();
+    QVector<QgsGeometry> resultCollection;
+    for ( QgsGeometry part : geomCollection )
     {
-      QgsGeometry newGeom( newGeometries.at( 0 ) );
-      newGeom.convertToMultiType();
+      QVector<QgsGeometry> newGeometries;
+      QgsPointSequence topologyTestPoints;
+      splitFunctionReturn = part.splitGeometry( splitLine, newGeometries, topologicalEditing, topologyTestPoints, false );
 
-      for ( int i = 1; i < newGeometries.size(); ++i )
+      if ( splitFunctionReturn == Qgis::GeometryOperationResult::Success && !newGeometries.isEmpty() )
       {
-        QgsGeometry part = newGeometries.at( i );
-        part.convertToSingleType();
-        newGeom.addPart( part );
-      }
-
-      mLayer->changeGeometry( feat.id(), newGeom );
+        for ( int i = 0; i < newGeometries.size(); ++i )
+        {
+          resultCollection.append( newGeometries.at( i ).asGeometryCollection() );
+        }
 
-      if ( topologicalEditing )
-      {
-        QgsPointSequence::const_iterator topol_it = topologyTestPoints.constBegin();
-        for ( ; topol_it != topologyTestPoints.constEnd(); ++topol_it )
+        if ( topologicalEditing )
         {
-          addTopologicalPoints( *topol_it );
+          QgsPointSequence::const_iterator topol_it = topologyTestPoints.constBegin();
+          for ( ; topol_it != topologyTestPoints.constEnd(); ++topol_it )
```

Comment:
I think it would be better to add the topology points after we call changeGeometry, especially if we return in the middle of the splitting operation (we could have topology points althought the split has not been effective

Reply:
Ok